### DISCUSSION TITLE
Enemy mob spawn conditions.

### DISCUSSION THREAD

**Discussion by sousuke0422** [Roadhog360/Et-Futurum-Requiem#152] [Ideas]
Same as implemented in gregtech 6.14.16.

`[CHANGED] Most Vanilla Hostile Mobs and Some Vanilla alike Hostiles, in the Overworld and Overworld alike Dimensions, will require a BLOCK Light Level of ZERO to spawn as opposed to 7 or less. This is to mirror a change I saw in one of the most recent experimental Snapshots, meaning Torches and other Light Sources have a much bigger range now. This does not affect Sky Light Levels.
`

**Comment by GregoriusT** [comment]
Relevant bit of Code to this:

```
	@SubscribeEvent
	public void onCheckSpawnEvent(LivingSpawnEvent.CheckSpawn aEvent) {
		if (aEvent.getResult() == Result.DENY) return;
		Class<? extends EntityLivingBase> aMobClass = aEvent.entityLiving.getClass();
		World aWorld = aEvent.world;
		int aX = UT.Code.roundDown(aEvent.x), aY = (int)UT.Code.bind(0, aWorld.getHeight(), UT.Code.roundDown(aEvent.y)), aZ = UT.Code.roundDown(aEvent.z);
		
		if (aWorld.provider.dimensionId != 0) return;
		if (SPAWN_HOSTILES_ONLY_IN_DARKNESS) try {
			Chunk tChunk = aWorld.getChunkFromBlockCoords(aX, aZ);
			if (tChunk != null && tChunk.getBlockStorageArray() != null && tChunk.getBlockStorageArray()[aY >> 4] != null && tChunk.getBlockStorageArray()[aY >> 4].getExtBlocklightValue(aX & 15, aY & 15, aZ & 15) > 0) {
				// Vanilla Mobs only, just in case.
				if (aMobClass == EntityCreeper.class || aMobClass == EntityEnderman.class || aMobClass == EntitySkeleton.class || aMobClass == EntityZombie.class || aMobClass == EntitySpider.class || aMobClass == EntityWitch.class || aMobClass == EntityBat.class) {aEvent.setResult(Result.DENY); return;}
				// Well, that Zombie is kindof like Vanilla, so it counts.
				if (MD.TC.mLoaded) if (aEvent.entityLiving instanceof EntityBrainyZombie) {aEvent.setResult(Result.DENY); return;}
				// TODO Add Drowned and other Et Futurum Requiem Mobs once they are released.
				if (MD.EtFu.mLoaded) if (aEvent.entityLiving instanceof EntityZombieVillager || aEvent.entityLiving instanceof EntityStray || aEvent.entityLiving instanceof EntityHusk) {aEvent.setResult(Result.DENY); return;}
			}
		} catch(Throwable e) {e.printStackTrace(ERR);}
	}
```

**Comment by Roadhog360** [comment]
> Relevant bit of Code to this:
> 
> ```
> 	@SubscribeEvent
> 	public void onCheckSpawnEvent(LivingSpawnEvent.CheckSpawn aEvent) {
> 		if (aEvent.getResult() == Result.DENY) return;
> 		Class<? extends EntityLivingBase> aMobClass = aEvent.entityLiving.getClass();
> 		World aWorld = aEvent.world;
> 		int aX = UT.Code.roundDown(aEvent.x), aY = (int)UT.Code.bind(0, aWorld.getHeight(), UT.Code.roundDown(aEvent.y)), aZ = UT.Code.roundDown(aEvent.z);
> 		
> 		if (aWorld.provider.dimensionId != 0) return;
> 		if (SPAWN_HOSTILES_ONLY_IN_DARKNESS) try {
> 			Chunk tChunk = aWorld.getChunkFromBlockCoords(aX, aZ);
> 			if (tChunk != null && tChunk.getBlockStorageArray() != null && tChunk.getBlockStorageArray()[aY >> 4] != null && tChunk.getBlockStorageArray()[aY >> 4].getExtBlocklightValue(aX & 15, aY & 15, aZ & 15) > 0) {
> 				// Vanilla Mobs only, just in case.
> 				if (aMobClass == EntityCreeper.class || aMobClass == EntityEnderman.class || aMobClass == EntitySkeleton.class || aMobClass == EntityZombie.class || aMobClass == EntitySpider.class || aMobClass == EntityWitch.class || aMobClass == EntityBat.class) {aEvent.setResult(Result.DENY); return;}
> 				// Well, that Zombie is kindof like Vanilla, so it counts.
> 				if (MD.TC.mLoaded) if (aEvent.entityLiving instanceof EntityBrainyZombie) {aEvent.setResult(Result.DENY); return;}
> 				// TODO Add Drowned and other Et Futurum Requiem Mobs once they are released.
> 				if (MD.EtFu.mLoaded) if (aEvent.entityLiving instanceof EntityZombieVillager || aEvent.entityLiving instanceof EntityStray || aEvent.entityLiving instanceof EntityHusk) {aEvent.setResult(Result.DENY); return;}
> 			}
> 		} catch(Throwable e) {e.printStackTrace(ERR);}
> 	}
> ```

Is this 1.7.10 code for the new light level 0 spawning logic?

**Comment by GregoriusT** [comment]
It is what GregTech-6 uses for it, and so far I did not see any bad Side Effects.

You DID remind me that I should probably check for some other Dimensions too though, like the Enviromine Cave Dimension. Edit: Aaaand I forgot I made the code snippet I gave you more simple with that dimension check, I already did before what I just got reminded of. XD

**Reply by Roadhog360** [reply]
Lol, I'll probably apply the new logic in the overworld only. Thanks for the code snippet.
